### PR TITLE
feat: Increase security by adding PipelineSecurityPolicies which enables a light in the tunnel for a multi-tenacy

### PR DESCRIPTION
**Pull request by keskad**
Introduces security-related enforcements on pipelines created by Lighthouse.
Resolves https://github.com/jenkins-x/jx/issues/8060

**How it works?**
- You can define an **OPTIONAL** policy that matches given repository or group of repositories. Policies are stored in central namespace in Kubernetes e.g. in `jx`
- If a pipeline is going to be spawned, then Lighthouse `security` module looks for policies, if there is a policy then it enforces it
- Enforcement of a policy can mean: `Changing a namespace`, `Changing a service account name` (in future it can also `Limit the pipeline max duration` or `Limit limit & request resources allocation`)
- The PipelineSecurityPolicies is a LAST DECISION before applying, it should overridde everything, especially things defined by non-privileged users
- Backward compatibility: NO POLICIES THEN NOTHING CHANGES. Policies are totally OPTIONAL.

**Edge cases:**
- There are policies, but regexp is not compillable. Then do not schedule any jobs. Because we want security policies, but they are invalid, we do not know what to allow
- Multiple policies matched for a Job. Then do not start this job, security must be defined to be **unambiguous and simple**

**Motivation:** 
- Allow to use Jenkins X in a large company that has tens of teams, and everything is security-driven, which means every team needs to be isolated. 
- Setting up JX per namespace was impossible due to hardcoded namespaces in various repositories and tools and was additionally too much overhead in terms of resources
- Policies are simple. 
- Running teams jobs in separated namespaces allows to restrict access to `cluster`, `network`, `git`, `registry`, `artifactory` by using namespaced policies and namespaced secrets

**Scope:**

Scope of this merge request is to add a **basic and optional feature** that allows to setup a cluster and work with for a client I'm working for. Support for things like `jx-pipeline-visualizer` would be a separate effort withing a separate PR, when basic things would allow mine project to start, then next steps could be possibly done if would be neccssary.

**Goal:** _Run jobs in different namespaces and using different service accounts, with working basic things like `gc-jobs`, with support for updating pull request statuses, chatops._

**Example of a security policy**

```yaml
---
apiVersion: lighthouse.jenkins.io/v1alpha1
kind: LighthousePipelineSecurityPolicy
metadata:
  name: keskad
  namespace: jx
spec:
  enforce:
    namespace: jx-damiank-cicd
    serviceAccountName: some-other-account
  repositoryPattern: keskad/springboot-(.*)
```

**Todo:** 
- [x] Unit tests
- [x] Documentation
- [x] Make sure gc jobs works for all jobs, including jobs spawned in separate namespaces
- [x] Make sure foghorn, keeper, poller are working with all jobs, including spawned in separate namespaces

**Testing space:**

Here I have my cluster configuration.

- https://github.com/keskad/jx3-test-charts
- https://github.com/keskad/jx3-local-testing-cluster

Linked PRs
----------------

- https://github.com/jenkins-x-plugins/jx-gitops/pull/835
- https://github.com/jenkins-x-plugins/jx-build-controller/pull/39
- https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/425
- https://github.com/jenkins-x/jx-helpers/pull/362 (no longer required as the solution changed, but having this improvement could be helpful in future)
- https://github.com/jenkins-x/jx-docs/pull/3560
- https://github.com/jenkins-x/jx-pipelines-visualizer/pull/145

### CODE REVIEW

**Comment by keskad** [comment]
image

- The first three bootjob entries are effect of testing policy for `keskad/(.*)` 
- 4th PR was intentionally failed by testing service account change: `translating TaskSpec to Pod: serviceaccounts "mietek-nie-dziala" not found`

**Comment by keskad** [comment]
Resolves https://github.com/jenkins-x/jx/issues/8060

File: Makefile [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Make does not offer a recursive wildcard function, so here's one:
 rwildcard=$(wildcard $1$2) $(foreach d,$(wildcard $1*),$(call rwildcard,$d/,$2))
 
+SHELL := /bin/bash
```
Comment (ankitm123) [inline Makefile:4]: Is this required? I use zsh, may be this will break it for non-bash users?

**Comment by ankitm123** [comment]
Looks great, thanks for this. Would be good to add some tests, so we dont introduce regression, and dont break things moving forward.

**Comment by keskad** [comment]
Created a Helm that is able to create secrets, service accounts, role etc. required for testing this PR. Later this could be moved to bootjob somehow or recommended in docs.

https://github.com/keskad/jx3-namespace-provision

**Comment by keskad** [comment]
Together with @yelhouti we were looking at options to simplify it a little bit.
There was an idea to create `LighthouseJobs` in `jx` and spawn `PipelineRuns` in various namespaces, but the owner references between those two objects were a blocker there, another option was making a cluster-scoped `LighthouseJob` but that would be a big breaking change, so sticked back to the original idea to spawn `LighthouseJob` in selected namespace.

**Comment by keskad** [comment]
What about roles for gc jobs? We have an idea that if somebody uses namespaced jobs, then in bootrepo there should be defined a list of namespaces, so the bootrepo can create `RoleBindings` for gc jobs for those namespaces to do not give `ClusterRoles` to jobs that can delete any pod :smile:

**Comment by keskad** [comment]
gc-activities
----------------

Changes to gc-activities are done as part of sub-PR: https://github.com/jenkins-x-plugins/jx-gitops/pull/835

**Comment by keskad** [comment]
jx-build-controller
-----------------------

There is a sub-PR: https://github.com/jenkins-x-plugins/jx-build-controller/pull/39 - it allows `jx-build-controller` to listen across all namespaces for JX-specific resources only.

**Comment by keskad** [comment]
jx-gitops
-----------

Support for multiple namespaces for `kind: CronJob` named `jx-gcactivities`

https://github.com/jenkins-x-plugins/jx-gitops/pull/835



jxboot-helmfile-resources
----------------------------------

sub-PR with changes required by `jx-gitops`

https://github.com/jenkins-x-charts/jxboot-helmfile-resources/pull/425

**Comment by yelhouti** [comment]
@keskad I commented on the jxboot-helmfile-resources charts

**Comment by keskad** [comment]
@yelhouti 

Thanks! Replied :sunglasses:

**Comment by keskad** [comment]
jx-pipelines-visualizer
-----------------------------

Added improvement to visualizer to be able to show pipelines from various namespaces: https://github.com/jenkins-x/jx-pipelines-visualizer/pull/145

**Comment by keskad** [comment]
Squashed and signed-off commits on my feature branch.

**Todo:**
- I will try those `ClusterRoles` + `RoleBindings` instead of `CR` + `CRB`
- Add support for `execution time limit` enforcement
- Unit tests

Then I remove "DRAFT" from this PR.

**Comment by keskad** [comment]
Ah, and also todo - changes in docs :sunglasses:

**Comment by keskad** [comment]
CR + CRB experiment
=================

**ClusterRole**

```yaml
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: lighthouse-tekton-controller-cluster-scoped
rules:
- apiGroups:
  - tekton.dev
  resources:
  - pipelines
  verbs:
  - get
- apiGroups:
  - tekton.dev
  resources:
  - pipelineruns
  verbs:
  - create
  - list
  - get
  - watch
- apiGroups:
  - lighthouse.jenkins.io
  resources:
  - lighthousebreakpoints
  - lighthousejobs
  verbs:
  - get
  - update
  - list
  - watch
- apiGroups:
  - lighthouse.jenkins.io
  resources:
  - lighthousejobs/status
  verbs:
  - update
  - get
  - watch
  - patch
```

**RoleBinding inside jx namespace**

```yaml
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: lighthouse-tekton-controller-experiment
  namespace: jx
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: lighthouse-tekton-controller-cluster-scoped
subjects:
- kind: ServiceAccount
  name: lighthouse-tekton-controller
  namespace: jx
```

**Result:**

```
E0309 15:11:31.783442       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.20.7/tools/cache/reflector.go:167: Failed to watch *v1alpha1.LighthouseBreakpoint: failed to list *v1alpha1.LighthouseBreakpoint: lighthousebreakpoints.lighthouse.jenkins.io is forbidden: User "system:serviceaccount:jx:lighthouse-tekton-controller" cannot list resource "lighthousebreakpoints" in API group "lighthouse.jenkins.io" at the cluster scope
```

**Comment by keskad** [comment]
@yelhouti Do you have please any hint? If we cant go this way, then we have I guess at least those options:

a) introduce label filtering in gc-pods at least, to match only pods of lighthouse and tekton [still: cronjob has wide permissions, that's bad]
b) schedule cronjob for selected namespaces? that could require more work - to do possibly
    b.1) in lighthouse helm chart (can a chart be installed across namespaces? there were an issue with that I remember)
    b.2) moving to other helm chart (would create a BC break for existing Jenkins X installations?)
    b.3) creating an operator that would copy cronjob from `jx` namespace to all namespaces labelled with `x=y`
    b.4) somehow propagated on bootjob level as a part of makefile? (rolebinding + cronjob) (may not work out of the box on existinig installatios of Jenkins X? - to check)

**Comment by ankitm123** [comment]
I will take a look today after work! :+1:

**Comment by keskad** [comment]
@ankitm123 Thanks, would be perfect. I would start then next day doing what you point out :wink:

**Comment by yelhouti** [comment]
Indeed unless we "fix" jx gitops, the only way I think this would work is by moving all the GC stuff to a separate helm chart and have it installed on all the namespaces where pipelines are run.

**Comment by keskad** [comment]
@yelhouti But moving to separate helm chart would not introduce a BREAKING CHANGE in existing Jenkins X installations? That would probably mean that people would need to add a new chart to their helmfiles in their gitops repos?

**Comment by keskad** [comment]
> Will just need to add the helm chart to the bootstrap repo helmfile, remove it for the jxboot-helmfile and explain that it should be added on other namespaces. (Including jx-git-operator where GC doesn't really work now)

Hmm... ok, but then I guess we cannot remove roles from Lightouse helm chart. We can only add a new helm chart only for separate namespaces?

**Comment by keskad** [comment]
To be enough clear for me, please let me know more about the separate helm chart concept.
What should it include? Should it have a `kind: Cronjob` + `ClusterRole` + `RoleBinding` and install per namespace or other way?  :slightly_smiling_face:

**Comment by yelhouti** [comment]
I don't understand why you say that.
I think we should remove everything related to GC (role, rolebinding, cronjob...) From jxboot-helmfile and move it to the new chart.
Am I missing something ?

**Comment by keskad** [comment]
Hmm... sure, extracting that from `jxboot-helmfile-resources` could be a clear solution. But what will be when someone upgrades Jenkins X, gc will disappear from  `jxboot-helmfile-resources` - will the Jenkins X upgrade mechanism add our new helm chart automatically?

Given we move to e.g. `jxboot-gcjobs` new helm chart those all gc jobs from `jx gitops` - how we could propagate `RoleBindings` to namespaces? Could we just add e.g. `jobsNamespaces` in helm values and create bindings for selected namespaces inside one helm or there will be a change required in bootrepo or somewhere else?

**Comment by keskad** [comment]
Added unit tests coverage, implemented maximum job duration enforcement as a third possible enforcement in the policy.

**Comment by keskad** [comment]
Related: https://github.com/jenkins-x/jx-helpers/pull/362 (required by sub-PR in jx-gitops)

**Comment by keskad** [comment]
Docs: https://github.com/jenkins-x/jx-docs/pull/3560

**Comment by keskad** [comment]
Today or tommorow I have to RETEST manually after recent changes to gc-jobs, then I will "undraft" the PR. 
I think that there will be no changes to lighthouse now from my side (except code review changes :sunglasses: ) , actually untested changes are in helmfile-bootrepo thing.

**Comment by keskad** [comment]
Ready for review @ankitm123

**Comment by keskad** [comment]
I reviewed this idea and have another one, closing for now.

**Pull request closed without merging**